plat/common/include/arm/arm64/paging.h

```diff
@@ -0,0 +1,342 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __PLAT_CMN_ARCH_PAGING_H__
+#error Do not include this header directly
+#endif
+
+#include <arm/arm64/cpu.h>
+#include <arm/arm64/cpu_defs.h>
+
+#include <uk/arch/limits.h>
+#include <uk/essentials.h>
+#include <uk/plat/paging.h>
+#include <uk/fallocbuddy.h>
+#include <uk/print.h>
+
+#include <errno.h>
+
+/* The implementation for arm64 supports a 48-bit virtual address space
+ * with 4KiB translation granule, as defined by VMSAv8.
+ *
+ * Notice:
+ * The Unikraft paging API uses the x86_64 convention, where pages are
+ * defined at L0, and the top level table is at L3. This is the opposite
+ * of the convention use by VMSAv8-A, where pages are defined at L3 and
+ * the top level table is defined at L0 (48-bit) / L-1 (52-bit). This
+ * implementation uses the Unikraft convention.
+ */
+
+/*
+ * We expect the physical memory to be mapped 1:1 into the following area in
+ * every virtual address space. This allows us to easily translate from virtual
+ * to physical page table addresses and vice versa. We also access the metadata
+ * for the frame allocators through this area.
+ */
+#define DIRECTMAP_AREA_START	0x0000ff8000000000
+#define DIRECTMAP_AREA_END	0xffffffffffffffff
+#define DIRECTMAP_AREA_SIZE	(DIRECTMAP_AREA_END - DIRECTMAP_AREA_START + 1)
+
+static inline __vaddr_t
+arm64_directmap_paddr_to_vaddr(__paddr_t paddr)
+{
+	UK_ASSERT(paddr < DIRECTMAP_AREA_SIZE);
+	return (__vaddr_t)paddr + DIRECTMAP_AREA_START;
+}
+
+static inline __paddr_t
+arm64_directmap_vaddr_to_paddr(__vaddr_t vaddr)
+{
+	UK_ASSERT(vaddr >= DIRECTMAP_AREA_START);
+	UK_ASSERT(vaddr < DIRECTMAP_AREA_END);
+	return (__paddr_t)(vaddr - DIRECTMAP_AREA_START);
+}
+
+static inline __pte_t
+pgarch_pte_create(__paddr_t paddr, unsigned long attr, unsigned int level,
+		  __pte_t template, unsigned int template_level __unused)
+{
+	__pte_t pte;
+
+	unsigned long pte_lx_map_paddr_mask[] = {
+		PTE_L0_PAGE_PADDR_MASK,
+		PTE_L1_BLOCK_PADDR_MASK,
+		PTE_L2_BLOCK_PADDR_MASK,
+	};
+
+	UK_ASSERT(PAGE_ALIGNED(paddr));
+	UK_ASSERT(level <= PT_MAP_LEVEL_MAX);
+
+	pte = paddr & pte_lx_map_paddr_mask[level];
+
+	if (!template)
+		pte |= ATTR_AF;
+	else
+		pte |= template & (ATTR_CONTIGUOUS | ATTR_DBM | ATTR_nG |
+				   ATTR_AF | ATTR_SH_MASK | ATTR_IDX_MASK);
+
+	if (level == PAGE_LEVEL)
+		pte |= Ln_PAGE;
+	else
+		pte |= Ln_BLOCK;
+
+	if (!(attr & PAGE_ATTR_PROT_WRITE))
+		pte |= ATTR_AP(ATTR_AP_RO);
+
+	if (!(attr & PAGE_ATTR_PROT_EXEC))
+		pte |= ATTR_XN;
+
+	switch (attr & (PAGE_ATTR_SHAREABLE_MASK <<
+		PAGE_ATTR_SHAREABLE_SHIFT)) {
+	case PAGE_ATTR_SHAREABLE_IS:
+		pte |= ATTR_SH(ATTR_SH_IS);
+		break;
+	case PAGE_ATTR_SHAREABLE_OS:
+		pte |= ATTR_SH(ATTR_SH_OS);
+		break;
+	case PAGE_ATTR_SHAREABLE_NS:
+		pte |= ATTR_SH(ATTR_SH_NS);
+		break;
+	default:
+		UK_CRASH("Invalid shareability type\n");
```

Comment:
This is pretty drastic but also justified as it can be expected that only valid flags should be supplied. However, considering this, I would do a `UK_ASSERT(0)` instead to save the code and data in release builds.

You can also try:
```suggestion
		UK_ASSERT(0 && "Invalid shareability type");
```